cmd/snap-confine/udev-support.c

```diff
@@ -18,301 +18,390 @@
-	if (udev_s->tagname_len == 0
-	    || udev_s->tagname_len >= MAX_BUF
-	    || strnlen(udev_s->tagname, MAX_BUF) != udev_s->tagname_len
-	    || udev_s->tagname[udev_s->tagname_len] != '\0')
-		die("snappy_udev->tagname has invalid length");
-
-	debug("%s: %s %s", __func__, path, udev_s->tagname);
-
-	struct udev_device *d =
-	    udev_device_new_from_syspath(udev_s->udev, path);
-	if (d == NULL)
-		die("cannot find device from syspath %s", path);
-	dev_t devnum = udev_device_get_devnum(d);
-	udev_device_unref(d);
-
-	_run_snappy_app_dev_add_majmin(udev_s, path, major(devnum),
-				       minor(devnum));
+	struct stat sbuf;
+
+	if (stat("/dev/uhid", &sbuf) == 0) {
+		sc_dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			   minor(sbuf.st_rdev));
+	}
 }
 
-/*
- * snappy_udev_init() - setup the snappy_udev structure. Return 0 if devices
- * are assigned, else return -1. Callers should use snappy_udev_cleanup() to
- * cleanup.
- */
-int snappy_udev_init(const char *security_tag, struct snappy_udev *udev_s)
+/**
+ * Allow access to /dev/net/tun
+ *
+ * When CONFIG_TUN=m, /dev/net/tun will exist but using it doesn't
+ * autoload the tun module but also /dev/net/tun isn't udev tagged
+ * until it is loaded. To work around this, if /dev/net/tun exists, add
+ * it unconditionally to the cgroup and rely on AppArmor to mediate the
+ * access. LP: #1859084
+ **/
+static void sc_udev_allow_dev_net_tun(int devices_allow_fd)
 {
-	debug("%s", __func__);
-	int rc = 0;
+	struct stat sbuf;
 
-	udev_s->tagname[0] = '\0';
-	udev_s->tagname_len = 0;
-	// TAG+="snap_<security tag>" (udev doesn't like '.' in the tag name)
-	udev_s->tagname_len = sc_must_snprintf(udev_s->tagname, MAX_BUF,
-					       "%s", security_tag);
-	for (size_t i = 0; i < udev_s->tagname_len; i++)
-		if (udev_s->tagname[i] == '.')
-			udev_s->tagname[i] = '_';
+	if (stat("/dev/net/tun", &sbuf) == 0) {
+		sc_dprintf(devices_allow_fd, "c %u:%u rwm", major(sbuf.st_rdev),
+			   minor(sbuf.st_rdev));
+	}
+}
 
-	udev_s->udev = udev_new();
-	if (udev_s->udev == NULL)
-		die("udev_new failed");
+/**
+ * Allow access to assigned devices.
+ *
+ * The snapd udev security backend uses udev rules to tag matching devices with
+ * tags corresponding to snap applications. Here we interrogate udev and allow
+ * access to all assigned devices.
+ **/
+static void sc_udev_allow_assigned(int devices_allow_fd, struct udev *udev,
+				   struct udev_list_entry *assigned)
+{
+	for (struct udev_list_entry * entry = assigned; entry != NULL;
+	     entry = udev_list_entry_get_next(entry)) {
+		const char *path = udev_list_entry_get_name(entry);
+		if (path == NULL) {
+			die("udev_list_entry_get_name failed");
+		}
+		struct udev_device *device =
+		    udev_device_new_from_syspath(udev, path);
+		if (device == NULL) {
+			die("cannot find device from syspath %s", path);
+		}
+		dev_t devnum = udev_device_get_devnum(device);
+		char type_c = strstr(path, "/block/") != NULL ? 'b' : 'c';
```

Comment:
Nitpick: `type_c` makes me think of 'c' as in 'character device'. Perhaps another name for the variable?

Not to mention, this doesn't contain the expected S_IFBLK/S_IFCHR switch from https://github.com/snapcore/snapd/commit/02b48d5645b0e4169dead186cb1b2a0a7e816b1e

Reply:
Done. Thanks for spotting this. The cost of leaving a branch open for so long.